### PR TITLE
[Organizations] Manage membership requests page

### PR DESCRIPTION
**Pull request by scottbommarito**
https://github.com/NuGet/NuGetGallery/issues/5550

This PR integrates the pending membership/transformation requests that a user has into their list of organizations.

image

### CODE REVIEW

File: src/NuGetGallery.Core/Entities/User.cs [hunk 1]
```diff
@@ -67,6 +67,11 @@ public User(string username)
         /// </summary>
         public virtual OrganizationMigrationRequest OrganizationMigrationRequest { get; set; }
 
+        /// <summary>
+        /// Requests for this user to become the admin of a <see cref="User"/> account that was transformed into an <see cref="Organization"/> account.
+        /// </summary>
+        public virtual ICollection<OrganizationMigrationRequest> OrganizationMigrationRequests { get; set; }
```
Comment (skofman1) [inline src/NuGetGallery.Core/Entities/User.cs]: >OrganizationMigrationRequests [](start = 65, length = 29)

Initialize with an empty list like we are doing for other collections.

File: src/NuGetGallery/Migrations/201803092333364_ConnectUsersToTheirOrganizationMigrationRequests.cs [hunk 2]
```diff
@@ -0,0 +1,16 @@
+namespace NuGetGallery.Migrations
+{
+    using System;
+    using System.Data.Entity.Migrations;
+    
+    public partial class ConnectUsersToTheirOrganizationMigrationRequests : DbMigration
+    {
+        public override void Up()
```
Comment (skofman1) [inline src/NuGetGallery/Migrations/201803092333364_ConnectUsersToTheirOrganizationMigrationRequests.cs]: It's an empty migration... why is it needed?

Comment (scottbommarito) [reply src/NuGetGallery/Migrations/201803092333364_ConnectUsersToTheirOrganizationMigrationRequests.cs]: I assumed it needed to create an index or something--will revert.

**Comment by skofman1** [comment]
>         public string Username { get; }

OrganizationName

---
Refers to: src/NuGetGallery/ViewModels/ManageOrganizationsItemViewModel.cs:10 in 65d72bd. [](commit_id = 65d72bd4dc42c6727070c429185325ba12291adf, deletion_comment = False)

**Comment by skofman1** [comment]
>         public string EmailAddress { get; }

Is e-mail address used?

---
Refers to: src/NuGetGallery/ViewModels/ManageOrganizationsItemViewModel.cs:11 in 65d72bd. [](commit_id = 65d72bd4dc42c6727070c429185325ba12291adf, deletion_comment = False)

File: src/NuGetGallery/ViewModels/ManageOrganizationsItemViewModel.cs [hunk 3]
```diff
@@ -12,15 +12,46 @@ public class ManageOrganizationsItemViewModel
         public bool CurrentUserIsAdmin { get; }
```
Comment (skofman1) [inline src/NuGetGallery/ViewModels/ManageOrganizationsItemViewModel.cs]: >CurrentUserIsAdmin [](start = 20, length = 18)

how is it used?

File: src/NuGetGallery/Views/Users/Organizations.cshtml [hunk 4]
```diff
@@ -59,9 +63,24 @@
...
+                                    else if (organization.Pending)
+                                    {
+                                        var accept = organization.Transform ? Url.ConfirmTransformAccount(organization.Username, organization.ConfirmationToken) : Url.AcceptOrganizationMembershipRequest(organization.Username, organization.ConfirmationToken);
+                                        var reject = organization.Transform ? Url.RejectTransformAccount(organization.Username, organization.ConfirmationToken) : Url.RejectOrganizationMembershipRequest(organization.Username, organization.ConfirmationToken);
+
+                                        <a class="btn" href="@accept" title="Accept Organization Membership Request" aria-label="@("Accept Organization Membership Request for " + organization.Username)">
+                                            <i class="ms-Icon ms-Icon--Accept" aria-hidden="true"></i>
+                                        </a>
+                                        <a class="btn" href="@reject" title="Reject Organization Membership Request" aria-label="@("Reject Organization Membership Request for " + organization.Username)">
```
Comment (skofman1) [inline src/NuGetGallery/Views/Users/Organizations.cshtml]: should stay on the same page.
We should use Ajax here to have a good experience. Please open am issue to track this improvement

File: src/NuGetGallery/Views/Users/Organizations.cshtml [hunk 5]
```diff
@@ -59,9 +63,24 @@
...
+                                            <i class="ms-Icon ms-Icon--Edit" aria-hidden="true"></i>
+                                        </a>
+                                    }
+                                    else if (organization.Pending)
+                                    {
+                                        var accept = organization.Transform ? Url.ConfirmTransformAccount(organization.Username, organization.ConfirmationToken) : Url.AcceptOrganizationMembershipRequest(organization.Username, organization.ConfirmationToken);
+                                        var reject = organization.Transform ? Url.RejectTransformAccount(organization.Username, organization.ConfirmationToken) : Url.RejectOrganizationMembershipRequest(organization.Username, organization.ConfirmationToken);
+
+                                        <a class="btn" href="@accept" title="Accept Organization Membership Request" aria-label="@("Accept Organization Membership Request for " + organization.Username)">
```
Comment (skofman1) [inline src/NuGetGallery/Views/Users/Organizations.cshtml]: stay on the same page for accept as well until we move to Ajax

File: src/NuGetGallery/ViewModels/ManageOrganizationsItemViewModel.cs (hunk 3 above)
Comment (scottbommarito) [reply src/NuGetGallery/ViewModels/ManageOrganizationsItemViewModel.cs]: used to show either admin/collab in UI

File: src/NuGetGallery/Views/Users/Organizations.cshtml (hunk 4 above)
Comment (scottbommarito) [reply src/NuGetGallery/Views/Users/Organizations.cshtml]: https://github.com/NuGet/NuGetGallery/issues/5669

File: src/NuGetGallery/Views/Users/Organizations.cshtml (hunk 5 above)
Comment (scottbommarito) [reply src/NuGetGallery/Views/Users/Organizations.cshtml]: https://github.com/NuGet/NuGetGallery/issues/5669

**Comment by scottbommarito** [comment]
@skofman1 `EmailAddress` is used for Gravatar and `CurrentUserIsAdmin` is used to show either "administrator" or "collaborator"

File: src/NuGetGallery/UrlExtensions.cs [hunk 6]
```diff
@@ -814,25 +814,34 @@ public static string AddOrganizationMember(this UrlHelper url, string accountNam
...
 
         public static string RejectOrganizationMembershipRequest(this UrlHelper url, MembershipRequest request, bool relativeUrl = true)
         {
-            return url.HandleOrganizationMembershipRequest("RejectMemberRequest", request, relativeUrl);
+            return url.RejectOrganizationMembershipRequest(request.Organization.Username, request.ConfirmationToken, relativeUrl);
         }
 
-        private static string HandleOrganizationMembershipRequest(
-            this UrlHelper url, string routeName, MembershipRequest request, bool relativeUrl = true)
+        public static string AcceptOrganizationMembershipRequest(this UrlHelper url, string organizationUsername, string confirmationToken, bool relativeUrl = true)
```
Comment (chenriksson) [inline src/NuGetGallery/UrlExtensions.cs]: Do you need both overloads of Accept/Reject?

File: src/NuGetGallery/ViewModels/ManageOrganizationsViewModel.cs [hunk 7]
```diff
@@ -12,7 +12,11 @@ public class ManageOrganizationsViewModel
 
         public ManageOrganizationsViewModel(User currentUser, IPackageService packageService)
         {
-            Organizations = currentUser.Organizations.Select(m => new ManageOrganizationsItemViewModel(m, packageService));
+            var organizations = currentUser.Organizations.Select(m => new ManageOrganizationsItemViewModel(m, packageService));
+            var pendingMemberships = currentUser.OrganizationRequests.Select(m => new ManageOrganizationsItemViewModel(m, packageService));
+            var pendingTransformations = currentUser.OrganizationMigrationRequests.Select(m => new ManageOrganizationsItemViewModel(m, packageService));
```
Comment (chenriksson) [inline src/NuGetGallery/ViewModels/ManageOrganizationsViewModel.cs]: This is requests to transform the user into an organization?

Why include it? Isn't there only 1 possible migration request, and doesn't that mean the user isn't a member of any other organizations?

File: src/NuGetGallery/ViewModels/ManageOrganizationsItemViewModel.cs [hunk 8]
```diff
@@ -7,20 +7,51 @@ namespace NuGetGallery
...
+        {
+            Pending = true;
+            ConfirmationToken = request.ConfirmationToken;
+        }
+
+        public ManageOrganizationsItemViewModel(OrganizationMigrationRequest request, IPackageService packageService)
+            : this(request.NewOrganization, true, packageService)
+        {
+            Pending = true;
+            Transform = true;
```
Comment (chenriksson) [inline src/NuGetGallery/ViewModels/ManageOrganizationsItemViewModel.cs]: What do Pending and Transform mean?

Comment (chenriksson) [reply src/NuGetGallery/ViewModels/ManageOrganizationsItemViewModel.cs]: Pending means current user is pending membership into this org?
Transform means current user has migration (to organization) request?

Could these be more descriptive?

File: src/NuGetGallery/Views/Users/Organizations.cshtml [hunk 9]
```diff
@@ -59,9 +63,24 @@
                                     @organization.PackagesCount
                                 </td>
                                 <td class="text-right align-middle package-controls">
-                                    <a class="btn" href="@Url.ManageMyOrganization(organization.Username)" title="Edit Organization" aria-label="@("Edit Organization " + organization.Username)">
-                                        <i class="ms-Icon ms-Icon--Edit" aria-hidden="true"></i>
-                                    </a>
+                                    @if (!organization.Pending && !organization.Transform)
```
Comment (chenriksson) [inline src/NuGetGallery/Views/Users/Organizations.cshtml]: Per comments above, if `Transform` indicates current user is being transformed into an organization... this shouldn't be possible, right?

File: src/NuGetGallery/ViewModels/ManageOrganizationsViewModel.cs [hunk 10]
```diff
@@ -12,7 +12,11 @@ public class ManageOrganizationsViewModel
...
+
+            Organizations = organizations.Concat(pendingMemberships).Concat(pendingTransformations);
```
Comment (chenriksson) [inline src/NuGetGallery/ViewModels/ManageOrganizationsViewModel.cs]: If the model code is getting more complex, I suggest extracting it into the Controller and adding unit tests.

File: src/NuGetGallery/ViewModels/ManageOrganizationsViewModel.cs (hunk 7 above)
Comment (scottbommarito) [reply src/NuGetGallery/ViewModels/ManageOrganizationsViewModel.cs]: no, `OrganizationMigrationRequests` is the list of requests with this user as admin

File: src/NuGetGallery/ViewModels/ManageOrganizationsItemViewModel.cs (hunk 8 above)
Comment (scottbommarito) [reply src/NuGetGallery/ViewModels/ManageOrganizationsItemViewModel.cs]: `Pending` = the user has a request to join this organization and is not actually in the organization
`Transform` = this pending request is for a user to become the admin of this organization when it is transformed

not sure how else to name these without being too descriptive. I can add some XML docs if that would help.

File: src/NuGetGallery/Views/Users/Organizations.cshtml (hunk 9 above)
Comment (scottbommarito) [reply src/NuGetGallery/Views/Users/Organizations.cshtml]: no, see comment above

File: src/NuGetGallery/ViewModels/ManageOrganizationsViewModel.cs (hunk 10 above)
Comment (scottbommarito) [reply src/NuGetGallery/ViewModels/ManageOrganizationsViewModel.cs]: I don't think it's very complex--it's just extracting some fields and concat-ing a set of models.

File: src/NuGetGallery/UrlExtensions.cs (hunk 6 above)
Comment (scottbommarito) [reply src/NuGetGallery/UrlExtensions.cs]: They are both used, one in the controller using the entities themselves and one in the view using the model.

**Comment by scottbommarito** [comment]
@skofman1 @chenriksson I've handled all feedback.

File: src/NuGetGallery/ViewModels/ManageOrganizationsItemViewModel.cs [hunk 11]
```diff
@@ -7,20 +7,51 @@ namespace NuGetGallery
...
+            : this(organization as User, isAdmin, packageService)
         {
-            var organization = membership.Organization;
-            Username = organization.Username;
+            OrganizationName = organization.Username;
             EmailAddress = organization.EmailAddress ?? organization.UnconfirmedEmailAddress;
-            CurrentUserIsAdmin = membership.IsAdmin;
+            CurrentUserIsAdmin = isAdmin;
             MemberCount = organization.Members.Count();
-            PackagesCount = packageService.FindPackageRegistrationsByOwner(membership.Organization).Count();
```
Comment (chenriksson) [inline src/NuGetGallery/ViewModels/ManageOrganizationsItemViewModel.cs]: We don't use `PackagesCount` for organizations?

File: src/NuGetGallery/ViewModels/ManageOrganizationsViewModel.cs (hunk 10 above)
Comment (chenriksson) [reply src/NuGetGallery/ViewModels/ManageOrganizationsViewModel.cs]: My point is that we don't always add unit tests for view models... maybe we should?

File: src/NuGetGallery/ViewModels/ManageOrganizationsItemViewModel.cs [hunk 12]
```diff
@@ -7,20 +7,51 @@ namespace NuGetGallery
...
+            IsPendingRequest = true;
+            ConfirmationToken = request.ConfirmationToken;
+        }
+
+        public ManageOrganizationsItemViewModel(OrganizationMigrationRequest request, IPackageService packageService)
+            : this(request.NewOrganization, true, packageService)
+        {
+            IsPendingRequest = true;
+            IsPendingTransformRequest = true;
+            ConfirmationToken = request.ConfirmationToken;
```
Comment (chenriksson) [inline src/NuGetGallery/ViewModels/ManageOrganizationsItemViewModel.cs]: Maybe it's useful to add comments for these constructors... the variations and differences between them is a bit confusing.

File: src/NuGetGallery.Core/Entities/EntitiesContext.cs [hunk 13]
```diff
@@ -162,6 +162,12 @@ protected override void OnModelCreating(DbModelBuilder modelBuilder)
                 .HasOptional(u => u.OrganizationMigrationRequest)
                 .WithRequired(m => m.NewOrganization);
 
+            modelBuilder.Entity<User>()
+                .HasMany(u => u.OrganizationMigrationRequests)
+                .WithRequired(m => m.AdminUser)
+                .HasForeignKey(m => m.AdminUserKey)
+                .WillCascadeOnDelete(true); // Migration request will be deleted with the Admin account.
```
Comment (chenriksson) [inline src/NuGetGallery.Core/Entities/EntitiesContext.cs]: No schema changes required for this, right?

File: src/NuGetGallery/UrlExtensions.cs [hunk 14]
```diff
@@ -1211,6 +1220,21 @@ public static string RejectTransformAccount(this UrlHelper url, User accountToTr
...
+
+        public static string ConfirmTransformAccount(this UrlHelper url, string accountToTransformUsername, string confirmationToken, bool relativeUrl = true)
+        {
+            return url.HandleTransformAccount(RouteName.TransformToOrganizationConfirmation, accountToTransformUsername, confirmationToken, relativeUrl);
+        }
+
+        public static string RejectTransformAccount(this UrlHelper url, string accountToTransformUsername, string confirmationToken, bool relativeUrl = true)
+        {
+            return url.HandleTransformAccount("RejectTransform", accountToTransformUsername, confirmationToken, relativeUrl);
+        }
```
Comment (chenriksson) [inline src/NuGetGallery/UrlExtensions.cs]: Add tests in UrlExtensionsFacts?

File: src/NuGetGallery/ViewModels/ManageOrganizationsItemViewModel.cs (hunk 12 above)
Comment (chenriksson) [reply src/NuGetGallery/ViewModels/ManageOrganizationsItemViewModel.cs]: Alternatively, maybe subclasses for variants?

File: src/NuGetGallery/ViewModels/ManageOrganizationsViewModel.cs (hunk 10 above)
Comment (scottbommarito) [reply src/NuGetGallery/ViewModels/ManageOrganizationsViewModel.cs]: I think we should if the behavior is complex...but I don't think it is here.

File: src/NuGetGallery.Core/Entities/EntitiesContext.cs (hunk 13 above)
Comment (scottbommarito) [reply src/NuGetGallery.Core/Entities/EntitiesContext.cs]: yup--had an empty migration in this PR earlier and didn't realize it until @skofman1 noticed it haha

File: src/NuGetGallery/UrlExtensions.cs (hunk 14 above)
Comment (scottbommarito) [reply src/NuGetGallery/UrlExtensions.cs]: We don't have tests for every new endpoint in `UrlExtensionsFacts`, just for `GetActionLink` itself (which is called by these endpoint functions).

File: src/NuGetGallery/ViewModels/ManageOrganizationsItemViewModel.cs (hunk 11 above)
Comment (scottbommarito) [reply src/NuGetGallery/ViewModels/ManageOrganizationsItemViewModel.cs]: This is already done by https://github.com/NuGet/NuGetGallery/pull/5665/files#diff-ab7f759b44970a4b0b7e813110774a34R24 so I removed it from here.

File: src/NuGetGallery/ViewModels/ManageOrganizationsItemViewModel.cs (hunk 12 above)
Comment (scottbommarito) [reply src/NuGetGallery/ViewModels/ManageOrganizationsItemViewModel.cs]: I think more comments would be helpful but I don't think subclasses would.

**Pull request merged**